context/src/main/java/io/micronaut/runtime/server/GracefulShutdownManager.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.runtime.server;
+
+import io.micronaut.context.annotation.Requires;
+import jakarta.inject.Singleton;
+
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Helper class that can be used to call all {@link GracefulShutdownLifecycle} beans.
+ *
+ * @since 4.5.0
+ * @author Jonas Konrad
+ */
+@Singleton
+@Requires(classes = GracefulShutdownLifecycle.class)
+public final class GracefulShutdownManager {
+    private final List<GracefulShutdownLifecycle> delegates;
+
+    GracefulShutdownManager(List<GracefulShutdownLifecycle> delegates) {
+        this.delegates = delegates;
+    }
+
+    /**
+     * Shut down all {@link GracefulShutdownLifecycle} beans. Semantics of this method are like
+     * {@link GracefulShutdownLifecycle#shutdownGracefully()}.
+     *
+     * @return A future that completes when all {@link GracefulShutdownLifecycle} beans have shut
+     * down
+     */
+    public CompletionStage<?> shutdownGracefully() {
```

Comment:
what calls `GracefulShutdownManager.shutdownGracefully()`?

Reply:
In this PR nothing. could make it a management endpoint, or just leave it to application code.

Reply:
there is already a shutdown endpoint so if enabled it should be called from that, but also we would likely need to integrate the shutdown hooks we add  already to handle SIGTERM

See https://dratler-shay.medium.com/how-to-optimize-graceful-shutdown-in-kubernetes-and-avoid-customer-impact-8c93ee3a6483

We need to be able to specify a termination grace period as well, in Spring this `spring.lifecycle.timeout-per-shutdown-phase=20s`